packages/app/features/groups/RoleFormScreen.tsx

```diff
@@ -0,0 +1,410 @@
+  useEffect(() => {
+    const unsubscribe = navigation.addListener('beforeRemove', (e) => {
+      // If no unsaved changes or already saving, let navigation proceed
+      if (!hasUnsavedChanges || isSavingRef.current) {
+        return;
+      }
+
+      // Prevent default behavior of leaving the screen
+      e.preventDefault();
+
+      // Store the navigation action and show confirm dialog
+      pendingNavigationAction.current = e.data.action;
+      setDiscardDialogOpen(true);
+    });
+
+    return unsubscribe;
+  }, [navigation, hasUnsavedChanges]);
+
+  const handleDiscardConfirm = useCallback(() => {
+    if (pendingNavigationAction.current) {
+      navigation.dispatch(pendingNavigationAction.current);
+      pendingNavigationAction.current = null;
+    }
+    setDiscardDialogOpen(false);
+  }, [navigation]);
+
+  const handleGoBack = useCallback(() => {
+    navigation.goBack();
+  }, [navigation]);
+
+  const rolesWithMembers = useMemo(() => {
+    if (!isEditMode) return [];
+    return groupRoles.filter((r) => {
+      return groupMembers.some((member) =>
+        member.roles.map((mr) => mr.roleId).includes(r.id!)
+      );
+    });
+  }, [isEditMode, groupRoles, groupMembers]);
+
+  const channelsCurrentlyInUse = useMemo(() => {
+    if (!isEditMode || !role) return [];
+    return groupChannels.filter((channel) => {
+      const writers = channel.writerRoles.map((r) => r.roleId);
+      const readers = channel.readerRoles.map((r) => r.roleId);
+      return writers.includes(role.id!) || readers.includes(role.id!);
+    });
+  }, [isEditMode, role, groupChannels]);
+
+  const handleSave = useCallback(
+    async (data: { title: string; description: string }) => {
+      if (!data.title) {
+        return;
+      }
+      isSavingRef.current = true;
+
+      if (isEditMode && role) {
+        // Edit existing role
+        const addMembers = selectedMembers.filter(
+          (id) => !initialMembers.includes(id)
+        );
+        const removeMembers = initialMembers.filter(
+          (id) => !selectedMembers.includes(id)
+        );
+
+        await updateGroupRole({
+          ...role,
+          title: data.title,
+          description: data.description,
+        });
+
+        // Add new members to the role
+        for (const contactId of addMembers) {
+          await addUserToRole(contactId, role.id!);
+        }
+        // Remove members from the role
+        for (const contactId of removeMembers) {
+          await removeUserFromRole(contactId, role.id!);
+        }
+      } else {
+        // Create new role
+        const newRoleId = generateSafeId(data.title, 'role');
+        await createGroupRole({
+          id: newRoleId,
+          title: data.title,
+          description: data.description,
+        });
+        // Add selected members to the role
+        for (const contactId of selectedMembers) {
+          await addUserToRole(contactId, newRoleId);
+        }
+      }
+
+      reset();
```

Comment:
If something goes wrong with the save the form gets reset and the user would need to just start over.

Try/catches with a toast alert and an ability to retry might be a good idea.